0000-abstract-return-types.md

```diff
@@ -0,0 +1,403 @@
+* In the `_dynamic` version using trait objects, the concrete argument type is
+  only known at runtime using a vtable.
+
+On the other hand, while you can write
+````rust
+fn produce_iter_dynamic() -> Box<Iterator<u8>>
+````
+you _cannot_ write something like
+````rust
+fn produce_iter_static() -> Iterator<u8>
+````
+That is, in today's Rust, abstract return types can only be written using trait objects, which
+can be a significant performance penalty. This RFC proposes "unboxed abstract
+types" as a way of achieving signatures like `produce_iter_static`. Like
+generics, unboxed abstract types guarantee static dispatch and inline data
+layout.
+
+Here are some problems that unboxed abstract types solve or mitigate:
+
+* _Returning unboxed closures_. The ongoing work on unboxed closures expresses
+  closures using traits. Sugar for closures generates an anonymous type
+  implementing a closure trait. Without unboxed abstract types, there is no way
+  to use this sugar while returning the resulting closure unboxed, because there
+  is no way to write the name of the generated type.
+
+* _Leaky APIs_. Functions can easily leak implementation details in their return
+  type, when the API should really only promise a trait bound. For example, a
+  function returning `Rev<Splits<'a, u8>>` is revealing exactly how the iterator
+  is constructed, when the function should only promise that it returns _some_
+  type implementing `Iterator<u8>`. Using newtypes/structs with private fields
+  helps, but is extra work. Unboxed abstract types make it as easy to promise only
+  a trait bound as it is to return a concrete type.
+
+* _Complex types_. Use of iterators in particular can lead to huge types:
+  ````rust
+  Chain<Map<'a, (int, u8), u16, Enumerate<Filter<'a, u8, vec::MoveItems<u8>>>>, SkipWhile<'a, u16, Map<'a, &u16, u16, slice::Items<u16>>>>
+  ````
+  Even when using newtypes to hide the details, the type still has to be written
+  out, which can be very painful. Unboxed abstract types only require writing the
+  trait bound.
+
+* _Documentation_. In today's Rust, reading the documentation for the `Iterator`
+  trait is needlessly difficult. Many of the methods return new iterators, but
+  currently each one returns a different type (`Chain`, `Zip`, `Map`, `Filter`,
+  etc), and it requires drilling down into each of these types to determine what
+  kind of iterator they produce.
+
+In short, unboxed abstract types make it easy for a function signature to
+promise nothing more than a trait bound, and do not generally require the
+function's author to write down the concrete type implementing the bound.
+
+# Detailed design
+
+> *Note*: this design borrows from
+>
+> * https://github.com/mozilla/rust/issues/11455,
+> * https://github.com/mozilla/rust/issues/10448 and
+> * https://github.com/mozilla/rust/issues/11196.
+
+The basic idea is to allow code like the following:
+````rust
+pub fn produce_iter_static() -> impl Iterator<int> {
```

Comment:
Maybe not the best place to write this, but the `impl` syntax sounds a bit confusing for me given the `impl Trait for Struct` syntax we currently have. What about (something like) the following?
```rust
pub fn produce_iter_static<I>() -> I guarantees I : Iterator<int> {
    range(0, 10).rev().map(|x| x * 2).skip(2)
}
```
This way, the usual syntax of static dispatch (the `where`-clause) is kept. 
As a bonus, it allows: 
```rust
pub fn produce_iter_static<I>() -> I guarantees I : Iterator<int> + Clone {...}
```